torch/csrc/jit/serialization/python_print.cpp

```diff
@@ -795,12 +800,47 @@ struct PythonPrintImpl {
     }
   }
 
+  static bool ContainsNonASCIIString(const IValue& val) {
+    if (val.isString()) {
+      const auto maxASCII = 0x7fu;
+      for (auto& c : val.toStringRef()) {
+        if (c > maxASCII) {
+          return true;
+        }
+      }
+    }
+
+    if (val.isList() || val.isTuple()) {
+      c10::ArrayRef<IValue> elems;
+      if (val.isTuple()) {
+        elems = val.toTuple()->elements();
+      } else {
+        elems = val.toListRef();
+      }
+      for (const auto& elem : elems) {
+        if (ContainsNonASCIIString(elem))
+          return true;
+      }
+    }
+
+    if (val.isGenericDict()) {
+      for (const auto& entry : val.toGenericDict()) {
+        if (ContainsNonASCIIString(entry.key()) ||
+            ContainsNonASCIIString(entry.value())) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
   void printConstant(TaggedStringStream& stmt, const IValue& v) {
     const auto customFormatter = [&](std::ostream& ss, const IValue& v) {
-      if (v.isTensor()) {
-        ss << "CONSTANTS.c" << getOrAddTensorConstant(v.toTensor());
+      if (v.isTensor() || ContainsNonASCIIString(v)) {
```

Comment:
This is a forward-compatibility breaking change, so it needs to be gated behind and if-statement and turned off for now, otherwise we might serialize a model with the new constants and older versions of the importer will error.

I recommend factoring out a new function (`shouldWriteToConstants` or something). For now, it must do the same as the old behavior. We can wait two weeks, then flip the switch to the new behavior.

Reply:
Done